vertical-pod-autoscaler/pkg/admission-controller/logic/recommendation_provider.go

```diff
@@ -31,104 +29,55 @@ import (
-	originalMilliRequest := big.NewInt(originalRequest.MilliValue())
-	originalMilliLimit := big.NewInt(originalLimit.MilliValue())
-	recommendedMilliRequest := big.NewInt(recommendedRequest.MilliValue())
-	var recommendedMilliLimit big.Int
-	recommendedMilliLimit.Mul(recommendedMilliRequest, originalMilliLimit)
-	recommendedMilliLimit.Div(&recommendedMilliLimit, originalMilliRequest)
-	if recommendedMilliLimit.IsInt64() {
-		return resource.NewMilliQuantity(recommendedMilliLimit.Int64(), recommendedRequest.Format), false
-	}
-	return resource.NewMilliQuantity(math.MaxInt64, recommendedRequest.Format), true
-}
-
 // GetContainersResources returns the recommended resources for each container in the given pod in the same order they are specified in the pod.Spec.
-func GetContainersResources(pod *v1.Pod, podRecommendation vpa_types.RecommendedPodResources, annotations vpa_api_util.ContainerToAnnotationsMap) []ContainerResources {
-	resources := make([]ContainerResources, len(pod.Spec.Containers))
+func GetContainersResources(pod *v1.Pod, podRecommendation vpa_types.RecommendedPodResources, limitRange *v1.LimitRangeItem,
+	annotations vpa_api_util.ContainerToAnnotationsMap) []vpa_api_util.ContainerResources {
+	resources := make([]vpa_api_util.ContainerResources, len(pod.Spec.Containers))
+	var defaultCpu, defaultMem, maxCpuLimit, maxMemLimit *resource.Quantity
+	if limitRange != nil {
+		defaultCpu = limitRange.Default.Cpu()
+		defaultMem = limitRange.Default.Memory()
+		maxCpuLimit = limitRange.Max.Cpu()
+		maxMemLimit = limitRange.Max.Memory()
+	}
 	for i, container := range pod.Spec.Containers {
-		resources[i] = newContainerResources()
-
 		recommendation := vpa_api_util.GetRecommendationForContainer(container.Name, &podRecommendation)
 		if recommendation == nil {
 			klog.V(2).Infof("no matching recommendation found for container %s", container.Name)
 			continue
 		}
-		resources[i].Requests = recommendation.Target
-
-		cpuLimit, capped := getProportionalLimit(container.Resources.Limits.Cpu(), container.Resources.Requests.Cpu(), resources[i].Requests.Cpu())
-		if cpuLimit != nil {
-			resources[i].Limits[v1.ResourceCPU] = *cpuLimit
-		}
-		if capped {
-			annotations[container.Name] = append(
-				annotations[container.Name],
-				fmt.Sprintf(
-					"Failed to keep CPU limit to request proportion of %d to %d with recommended request of %d milliCPU; doesn't fit in int64. Capping limit to MaxInt64",
-					container.Resources.Limits.Cpu().MilliValue(), container.Resources.Requests.Cpu().MilliValue(), resources[i].Requests.Cpu().MilliValue()))
-		}
-		memLimit, capped := getProportionalLimit(container.Resources.Limits.Memory(), container.Resources.Requests.Memory(), resources[i].Requests.Memory())
-		if memLimit != nil {
-			resources[i].Limits[v1.ResourceMemory] = *memLimit
+		cpuLimit, annotation := vpa_api_util.GetProportionalLimit(container.Resources.Limits.Cpu(), container.Resources.Requests.Cpu(), recommendation.Target.Cpu(), defaultCpu)
```

Comment:
I was hoping to have a list of strings returned and lose the if, but I guess this is fine as well.